lib/tree.rb

```diff
@@ -2,59 +2,168 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
+
+  def inorder(array)
+    #left, root, right
+    if @left != nil
+      @left.inorder(array)
+    end
+
+    array << {key: @key, value: @value}
+
+    if @right != nil
+      @right.inorder(array)
+    end
+  end
+
+  def preorder(array)
+    #Root, Left, Right
+    array << {key: @key, value: @value}
+
+    if @left != nil
+      @left.preorder(array)
+    end
+
+    if @right != nil
+      @right.preorder(array)
+    end
+  end
+
+  def postorder(array)
+    #left, right, root
+
+    if @left != nil
+      @left.postorder(array)
+    end
+
+    if @right != nil
+      @right.postorder(array)
+    end
+
+    array << {key: @key, value: @value}
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n) if it is balanced, O(n) if it is not
+  # Space Complexity: Constant
```

Comment:
Actually since `add_helper` is recursive the space complexity will be O(log n) if the tree is balanced and O(n) if it's not due to the function call stack.